### PR TITLE
E2E/Playwright: Add testcontainers to playwright-lib

### PR DESCRIPTION
**Pull request by saturninoabril**
#### Summary

Replaces the bash/Compose layer that provisions the server + dependency stack for Playwright with [Testcontainers](https://node.testcontainers.org/), colocated with the test code.

**Two deployment modes** (`lib/src/test_config.ts`):
- **`external` (default, outside Playwright codebase)** — current behavior, unchanged: still default to `localhost:8065` with `PW_BASE_URL`override. This is either run from source (`server/Makefile` / `make run`) or from container (`e2e-tests/Makefile` / `make`)
- **`testcontainers` (with `PW_USE_TESTCONTAINERS=true`, within Playwright codebase)** — brings up test server via Testcontainers then tears everything down after the run.

**New container orchestration** (`lib/src/containers/`):
- `stack.ts` — start/stop orchestration, container reuse (`PW_TESTCONTAINERS_REUSE`), env-file handoff to worker processes (`.env.testcontainers`), and adoption of an already-running stack.
- `network.ts`, `requirements.ts`, `env_baseline.ts`, `constants.ts`, `default_images.ts`, `paths.ts` — shared plumbing.
- Per-service container modules: `postgres_container.ts`, `inbucket_container.ts`, `mattermost_container.ts`, `openldap_container.ts`, `keycloak_container.ts`, `elasticsearch_container.ts`, `opensearch_container.ts`, `minio_container.ts`, `azurite_container.ts`, `mmctl_container.ts`, `webhook_container.ts`.

Test steps:
- `cd e2e-tests/playwright && npm run test` — unchanged (`external` mode, default).
- `cd e2e-tests/playwright && PW_USE_TESTCONTAINERS=true npm run test -- some.spec.ts` — verify the stack comes up standalone, without a separately-running server.
- `PW_USE_TESTCONTAINERS=true PW_TESTCONTAINERS_SERVICES=openldap,keycloak,elasticsearch,opensearch,minio,azurite npm run test -- specs/functional/system_console/{ldap,saml,search,file_storage,mmctl}/**` — verify each additional-service spec passes against a real container.
- `npm run testcontainers:down` — verify it archives `.env.testcontainers` and removes containers/network even with nothing currently running.

Tests passed in test-system-io staging env using Testcontainers ([one of the test runs in staging](https://staging-test-io.test.mattermost.com/reports/mattermost/pr-37570/0488119/playwright-testcontainers-dispatch-test?gh_run_id=30238142226&gh_run_attempt=1))
Screenshot 2026-07-27 at 9 24 21 AM

Tests passed in test-system-io prod env using existing docker-compose (no breaking change, existing setup works as is, [one of the test runs in prod](https://test-io.test.mattermost.com/reports/mattermost/pr-37570/0488119/playwright-full-enterprise?gh_run_id=30236504641&gh_run_attempt=1))
Screenshot 2026-07-27 at 9 26 14 AM

#### Ticket Link
none

#### Release Note
```release-note
NONE
```

## Change Impact: 🟡 Medium

**Regression Risk:** The change is largely confined to the Playwright/Testcontainers E2E harness but it heavily refactors shared test infrastructure (global stack lifecycle with adoption/reuse, networking/addressing via internalBaseURL/webhookInternalUrl, and teardown/log archival), which can introduce flakiness if any container/network/env assumptions differ between run modes. It also expands coverage around external integrations (webhook/OAuth/mmctl) and file backend verification (object-key existence), so mis-wired URLs or object-key checks could break multiple test suites at once.

**QA Recommendation:** Keep manual QA minimal; rely on CI with full/Testcontainers + reuse enabled, and add targeted validation for: permalink/ABAC edge cases, mock file server reachability (SVG/link previews), webhook integration URL targeting, OAuth/mmctl remote execution, and file upload/object-key assertions plus `testcontainers:down` cleanup/log collection. A quick local smoke run of one or two representative specs is enough if you frequently touch container lifecycle code.

*Generated by CodeRabbitAI*

### CODE REVIEW

File: e2e-tests/playwright/lib/src/server/keycloak.ts [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+    const {adminClient} = await getAdminClient();
+
+    try {
+        const config = await samlServerConfig(adminClient);
+        await adminClient.patchConfig({SamlSettings: config});
+    } catch (error) {
+        test.skip(true, `Skipping test - Keycloak SAML setup failed: ${String(error)}`);
+    }
+}
```
[the remark this answers is not in the document]
Comment (saturninoabril) [reply e2e-tests/playwright/lib/src/server/keycloak.ts:151]: adminClient is required for testing and so correct to fail hard than skip.

File: e2e-tests/playwright/lib/src/containers/paths.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+// Resolved relative to this module's own location (not the caller's cwd or the Mattermost
+// monorepo), so it keeps working once `@mattermost/playwright-lib` is installed as an npm
+// package with no access to the rest of the repo. `preserveModules` keeps `containers/assets`
+// alongside the compiled output of this file in `dist`, same as `src`.
+const assetsDir = path.join(__dirname, 'assets');
+
+export function containerAssetPath(...segments: string[]): string {
+    return path.join(assetsDir, ...segments);
+}
```
[the remark this answers is not in the document]
Comment (saturninoabril) [reply e2e-tests/playwright/lib/src/containers/paths.ts:19]: Should be fine

[the remark this answers is not in the document]
Comment (saturninoabril) [reply e2e-tests/playwright/lib/src/containers/paths.ts:19]: Remain as is

File: e2e-tests/playwright/global_setup.ts [hunk 3]
```diff
@@ -1,21 +1,26 @@
...
         // eslint-disable-next-line no-console
         console.error(error);
-        throw new Error(
-            `Global setup failed.\n\tEnsure the server at ${testConfig.baseURL} is running and accessible.\n\tPlease check the logs for more details.`,
-        );
+        const message = error instanceof Error ? error.message : String(error);
+        const hint = testConfig.useTestContainers
+            ? 'Check the container named above and its logs under logs/.'
+            : `Ensure the server at ${testConfig.baseURL} is running and accessible.`;
+        throw new Error(`Global setup failed: ${message}\n\t${hint}`, {cause: error});
```
Comment (yasserfaraazkhan) [inline e2e-tests/playwright/global_setup.ts]: we should add await stopStack() before throwing
so that if the startstack has started few services and throws error, we can stop them and then throw error

**Review by yasserfaraazkhan — APPROVED** [review]

File: e2e-tests/playwright/global_setup.ts (hunk 3 above)
Comment (saturninoabril) [reply e2e-tests/playwright/global_setup.ts]: Good point, applied, thanks.

**Comment by saturninoabril** [comment]
@amyblais I'll be cherry-picking this to supported releases

**Comment by amyblais** [comment]
> I'll be cherry-picking this to supported releases

Ok 👍

**Comment by saturninoabril** [comment]
Hi @amyblais, likely we don't need with https://github.com/mattermost/mattermost/pull/37697 now merged, but I'll keep on watching.

**Comment by amyblais** [comment]
> likely we don't need with #37697 now merged, but I'll keep on watching.

Ok 👍
